controllers/agentconfig_controller.go

```diff
@@ -0,0 +1,706 @@
+	lables := agentCfg.Spec.GetPluginsLabels()
+	lables[porterv1.LabelResourceName] = agentCfg.Name
+
+	selector := &metav1.LabelSelector{
+		MatchLabels: lables,
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      agentCfg.GetPVCName(),
+			Namespace: agentCfg.Namespace,
+			Labels:    agentCfg.Spec.GetPluginsLabels(),
+			OwnerReferences: []metav1.OwnerReference{
+				{ // I'm not using controllerutil.SetControllerReference because I can't track down why that throws a panic when running our tests
+					APIVersion:         agentCfg.APIVersion,
+					Kind:               agentCfg.Kind,
+					Name:               agentCfg.Name,
+					UID:                agentCfg.UID,
+					Controller:         pointer.BoolPtr(true),
+					BlockOwnerDeletion: pointer.BoolPtr(true),
+				},
+			},
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadOnlyMany},
+			Selector:    selector,
+			Resources: corev1.ResourceRequirements{
+				Requests: map[corev1.ResourceName]resource.Quantity{
+					corev1.ResourceStorage: agentCfg.Spec.GetVolumeSize(),
+				},
+			},
+		},
+	}
+
+	if err := r.Create(ctx, pvc); err != nil {
+		return nil, errors.Wrap(err, "error creating the agent volume (pvc)")
+	}
+
+	return pvc, nil
+
+}
+
+// removeAgentCfgFinalizer deletes the porter finalizer from the specified resource and saves it.
+func removeAgentCfgFinalizer(ctx context.Context, log logr.Logger, client client.Client, agentCfg *porterv1.AgentConfig) error {
+	log.V(Log5Trace).Info("removing finalizer")
+	controllerutil.RemoveFinalizer(agentCfg, porterv1.FinalizerName)
+	return client.Update(ctx, agentCfg)
+}
+
+func (r *AgentConfigReconciler) shouldDelete(agentCfg *porterv1.AgentConfig) bool {
+	// ignore a deleted CRD with no finalizers
+	return isDeleted(agentCfg) && isFinalizerSet(agentCfg)
+}
+
+// cleanup remove the owner references on both pvc and pv so the when no resource is referencing them, GC can clean them up after the agentCfg has been deleted
+func (r *AgentConfigReconciler) cleanup(ctx context.Context, log logr.Logger, agentCfg *porterv1.AgentConfig) error {
+	hashPVC := agentCfg.GetPVCName()
+	log.V(Log4Debug).Info("Start cleaning up persistent volume claim.", "persistentvolumeclaim", hashPVC, "namespace", agentCfg.Namespace)
+	key := client.ObjectKey{Namespace: agentCfg.Namespace, Name: hashPVC}
+	newPVC := &corev1.PersistentVolumeClaim{}
+	err := r.Get(ctx, key, newPVC)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	pv := &corev1.PersistentVolume{}
+	pvKey := client.ObjectKey{Namespace: newPVC.Namespace, Name: newPVC.Spec.VolumeName}
+	err = r.Get(ctx, pvKey, pv)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if idx, exist := containOwner(pv.GetOwnerReferences(), agentCfg); exist {
```

Comment:
Add comment here and below on what's different between these two calls to removeOwnerByIdx